flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java

```diff
@@ -60,12 +60,17 @@ protected void assignPartitionsToConsumer(KafkaConsumer<byte[], byte[]> consumer
 		consumer.assign(topicPartitions);
 	}
 
+	@Override
+	protected void emitRecord(T record, KafkaTopicPartitionState<TopicPartition> partition, long offset, ConsumerRecord consumerRecord) throws Exception {
+		// pass timestamp
+		super.emitRecord(record, partition, offset, consumerRecord.timestamp());
+	}
+
 	/**
 	 * Emit record Kafka-timestamp aware.
 	 */
 	@Override
-	protected <R> void emitRecord(T record, KafkaTopicPartitionState<TopicPartition> partitionState, long offset, R kafkaRecord) throws Exception {
-		long timestamp = ((ConsumerRecord) kafkaRecord).timestamp();
+	protected void emitRecord(T record, KafkaTopicPartitionState<TopicPartition> partitionState, long offset, long timestamp) throws Exception {
 		if (timestampWatermarkMode == NO_TIMESTAMPS_WATERMARKS) {
```

Comment:
Is it possible to let `AbstractFetcher#emitRecord` determine whether to call `collectWithTimestamp` or `collect` depending on the provided timestamp (== `Long.MIN_VALUE`)? Then, we won't need to override the base `emitRecord` here, correct?